feat: full support for google credentials in gcloud-aio clients

m1racoli opened the pull request:
The class CredentialsToken implements the ability to generate access tokens to be used in gcloud-aio clients from Google credentials objects provided by instances of Google Cloud hooks.

With this change we provide all credentials based capabilities of Google Cloud hooks (for example impersonation) to `gcloud-aio` clients.

closes: #34727
related: #36341 

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

Commit:
feat: full support for google credentials in gcloud-aio clients

The class CredentialsToken implements the ability to generate access
tokens to be used in gcloud-aio clients from Google credentials objects
provided by instances of Google Cloud hooks.

With this change we provide all credentials based capabilities of Google
Cloud hooks (for exmaple impersonation) to gcloud-aio clients.

m1racoli committed:
Merge branch 'main' into gcloud-aio-impersonation

m1racoli committed:
Merge branch 'main' into gcloud-aio-impersonation

Commit:
test: add tests for CredentialsToken

Lee-W commented on airflow/providers/google/cloud/hooks/bigquery.py:
```diff
@@ -3513,11 +3518,11 @@ async def get_table_client(
-                dataset_name=dataset,
-                table_name=table_id,
-                project=project_id,
-                service_file=file,
```
Will this affected how `service_files` was called?

m1racoli replied:
Service file will be ignored and is no longer needed with our custom `CredentialsToken` implementation, which instead generates the access token from a given credentials object.

Lee-W commented on airflow/providers/google/common/hooks/base_google.py:
```diff
@@ -623,6 +628,51 @@ def test_connection(self):
+            session=session,
+        )
+
+    async def get_project(self) -> str | None:
+        return self.project
```
May I know what is this method for? I do not see it's being used elsewhere

m1racoli replied:
We are subclassing `cloud.aio.auth.token.Token` which is used in the corresponding `gcloud-aio` clients. It mostly provides two asynchronous methods `get` and `get_project`, which are used in the clients to perform certain actions. Without it the implementation of `CredentialsToken` would be incomplete and could cause problems.

Even though we currently pass the `project` argument in all client instances, not re-implementing this method could cause side effects in those clients in the future. All clients rely on the [token's get_project](https://github.com/talkiq/gcloud-aio/blob/f1132b005ba35d8059229a9ca88b90f31f77456d/bigquery/gcloud/aio/bigquery/bigquery.py#L93) method, if no project has been specified.

m1racoli replied:
I want to add, even if we pass the project to the clients, if the project is `None` then the clients rely on the project provided by the token instance, which itself relies on the hook's `get_credentials_and_project_id`. Which should be the desired outcome.

Lee-W commented on tests/providers/google/common/hooks/test_base_google.py:
```diff
@@ -874,3 +875,95 @@ def test_should_fallback_when_empty_string_in_env_var(self):
+        mock_credentials.refresh.assert_called_once()
+
+    @pytest.mark.asyncio
+    @mock.patch(MODULE_NAME + ".get_credentials_and_project_id", return_value=("CREDENTIALS", "PROJECT_ID"))
```
```suggestion
    @mock.patch(f"{MODULE_NAME}.get_credentials_and_project_id", return_value=("CREDENTIALS", "PROJECT_ID"))
```

Lee-W reviewed (commented):
Overall, it looks great! Just left a few minor comments

m1racoli committed:
Update tests/providers/google/common/hooks/test_base_google.py

Co-authored-by: Wei Lee <weilee.rx@gmail.com>

pankajastro commented on airflow/providers/google/common/hooks/base_google.py:
```diff
@@ -639,6 +689,7 @@ async def get_sync_hook(self) -> Any:
             self._sync_hook = await sync_to_async(self.sync_hook_class)(**self._hook_kwargs)
         return self._sync_hook
 
-    async def service_file_as_context(self) -> Any:
```
I'll call it a breaking change. We will have to either restore this method and add a deprecation for future removal or this PR can go in a major release so need to update the changelog accordingly

m1racoli replied:
You are right! I only checked the use of this method within the airflow repo, but indeed it's public.

I can keep it in. Doesn't necessarily need to be deprecated though.

In this context I will also make CredentialsToken non-public, to avoid issues in the future.

m1racoli replied:
@pankajastro I've added some changes to address your comment.

pankajastro reviewed (approved):
Just have one minor comment https://github.com/apache/airflow/pull/36849/files#r1463026613 otherwise LGTM

Commit:
Merge remote-tracking branch 'upstream/main' into gcloud-aio-impersonation

Commit:
refactor: make CredentialsToken private

This class is only intended to be used within the Google provider and
might need to change in the future. Making it private in order to avoid
a potential breaking change in the future.

Commit:
Revert removal of service_file_as_context

The method `service_file_as_context` not being used anymore in the
airflow, but it is public and removing would imply a breaking changes
for users for the Google provider. Therefore we keep it.
